Add a combined point + line AGOL layer

johnclary opened the pull request:
## Associated issues
- https://github.com/cityofaustin/atd-data-tech/issues/14477


## Testing
**URL to test:** Local

**Steps to test:**
1. Start moped with a fresh copy of production data
2. Follow the steps to run the ETL: https://github.com/cityofaustin/atd-moped/pull/1148#issue-1927133588

Check out the results in the [AGOL dataset](https://austin.maps.arcgis.com/home/item.html?id=1c084c8756a84e6db7e2796c98c850a2), which now has a 3rd layer for the combined features.

<img width="362" alt="Screenshot 2023-11-03 at 12 36 50 PM" src="https://github.com/cityofaustin/atd-moped/assets/14793120/57059f33-6a5e-4b8d-a9be-ae43fbf447ab">


---
#### Ship list
- [ ] Code reviewed 
- [ ] Product manager approved
- [ ] Added to [QA test script](https://docs.google.com/spreadsheets/d/1n_O6MLh9cwwPf57HUM394Ea-z9uuoEV1-QW4axNZXLE/edit#) if applicable

johnclary commented on moped-database/metadata/tables.yaml:
note to self: need to add the `moped_components.line_representation` column to this view, so that it's possible to know which features in the combined agol layer are line-ified points

johnclary commented on moped-database/views/component_arcgis_online_view.sql:
```diff
@@ -1,176 +1,208 @@
+        comp_geography.length_feet_total, 
+        mc.component_name, 
+        mc.component_subtype, 
+        mc.component_name_full, 
```
using the column Chia added

johnclary commented on moped-database/views/component_arcgis_online_view.sql:
```diff
@@ -1,176 +1,208 @@
+        comp_geography.feature_ids, 
+        mpc.component_id, 
+        comp_geography.geometry, 
+        comp_geography.line_geometry, 
```
here's the new `line_geometry` column

johnclary commented on moped-database/views/component_arcgis_online_view.sql:
i'm sorry that i screwed up this diff so badly :/

here is a useful diff: https://www.diffchecker.com/75UwnR81/

johnclary replied:
we need to figure out a SQL prettifier for the team 😑

chiaberry replied:
the diff link doesnt work for me, it says it doesnt exist

mddilley replied:
+1 about a SQL formatter. I want to try [this vscode plugin](https://marketplace.visualstudio.com/items?itemName=ReneSaarsoo.sql-formatter-vsc). It is the official plugin for [sql-formatter](https://www.npmjs.com/package/sql-formatter) which looks like a well-maintained library. Downside is there is no CLI. 🫤

johnclary replied:
neat—i will check it out.

johnclary commented on moped-database/views/component_arcgis_online_view.sql:
```diff
@@ -1,176 +1,208 @@
+                id, 
+                feature_street_segments.component_id, 
+                feature_street_segments.geography :: geometry, 
+                feature_street_segments.geography :: geometry as line_geography, 
```
for features that are already lines we just re-use the `geometry` column

johnclary commented on moped-etl/arcgis/components_to_agol.py:
```diff
@@ -65,19 +64,49 @@ def main():
+            line_feature = make_esri_feature(
+                esri_geometry_key="paths", geometry=line_geometry, attributes=component
+            )
+            all_features["combined"].append(line_feature)
```
this looks like more refactoring than it really was bc i tweaked some awkward function signatures. tldr: we use the `line_geomtery` column to create a third group of features that is the superset of points + lines.

mddilley reviewed (approved):
Code looks great, and thanks for the comments. 🙏 I kept getting random 404s from the AGOL api when running the script, but it works and I can see all three layers when I explored the linked dataset. 🗺️ ✨

johnclary commented:
i am pretty concerned that you both hit API errors. i guess we can just ship this and see how big of a problem it is 😞 

in knack services we have this ["resilient" error handler](https://github.com/cityofaustin/atd-knack-services/blob/6bf34e4810e7767c07940dc9ffe5108ac6e969d4/services/records_to_agol.py#L33) to combat AGOL API errors. it doesn't sound like it would have addresses the issues y'all encountered though...

mddilley commented:
@johnclary I kind of wrote it off because I've had similar problems with the Socrata API locally and that DAG ran without a hitch. Sounds like this is different though so here are the errors that I ran into if it is helpful. I just realized that it was two different messages coming back.

**404 error**
```
Traceback (most recent call last):
  File "/app/components_to_agol.py", line 126, in <module>
    main()
  File "/app/components_to_agol.py", line 121, in main
    add_features(feature_type, feature_chunk)
  File "/app/utils.py", line 137, in add_features
    res.raise_for_status()
  File "/usr/local/lib/python3.11/site-packages/requests/models.py", line 1021, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://services.arcgis.com/0L95CJ0VTaxqcmED/ArcGIS/rest/services/Moped_Project_Components/FeatureServer/0/addFeatures
```

**Token error**
```
Traceback (most recent call last):
  File "/app/components_to_agol.py", line 126, in <module>
    main()
  File "/app/components_to_agol.py", line 121, in main
    add_features(feature_type, feature_chunk)
  File "/app/utils.py", line 139, in add_features
    handle_arcgis_response(responseData)
  File "/app/utils.py", line 79, in handle_arcgis_response
    raise ValueError(responseData["error"])
ValueError: {'code': 499, 'message': 'Token Required', 'messageCode': 'GWM_0003', 'details': ['Token Required']}
```

johnclary commented:
ah—thanks for that. those errors make no sense to me! e.g. we're making sure you have a token 🤷 

```python
    try:
        os.environ["AGOL_TOKEN"] = res.json()["token"]
    except Exception as e:
        raise ValueError(f"Unable to get token: {e.response.text}")
```

we'll see how this does when it's deployed 😬
